pkg/frontend/querymiddleware/codec.go

```diff
@@ -112,7 +116,46 @@ type Response interface {
 	GetHeaders() []*PrometheusResponseHeader
 }
 
-type prometheusCodec struct{}
+type prometheusCodecMetrics struct {
+	duration *prometheus.HistogramVec
+	size     *prometheus.HistogramVec
+}
+
+func newPrometheusCodecMetrics(registerer prometheus.Registerer) *prometheusCodecMetrics {
+	factory := promauto.With(registerer)
+	ms := 1.0 / 1000
+	us := ms / 1000
+	ns := us / 1000
+	kb := 1024.0
+	mb := 1024 * kb
+
+	return &prometheusCodecMetrics{
+		duration: factory.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: "cortex",
+			Name:      "frontend_query_payload_codec_duration_seconds",
+			Help:      "Total time spent encoding or decoding query result payloads, in seconds.",
+			// We need a wide range of buckets here because encoding protobuf payloads can take as little as 80ns, and decoding very large JSON payloads can take many milliseconds.
+			Buckets: []float64{50 * ns, 100 * ns, 500 * ns, 1 * us, 5 * us, 10 * us, 25 * us, 50 * us, 100 * us, 250 * us, 500 * us, 1 * ms, 2.5 * ms, 5 * ms, 10 * ms, 20 * ms},
```

Comment:
This seems like quite a lot of buckets very close in value. Is there value in differentiating between these cases at this granularity? 

I would have expected something closer to

* minimum of `1us` - anything lower than this is "fast enough to not care"
* `100us`  - noticeable amount of time to spend serializing something 
* `1ms` - significant overhead
* `25ms` - something is odd with this request

Reply:
Fair point, I've changed this in 44d06cf.

Reply:
(I can't wait for native histograms where I won't have to think so much about this stuff)